src/queryBuilder.ts

```diff
@@ -0,0 +1,197 @@
+ * @typeParam T - The type of the object to query. This is generally not needed to be specified, but can be useful if you're calling this yourself instead of through a {@link Model}.
+ */
+export type GenerateQueryOptions<T extends object> = {
+	where?: Partial<T>;
+	limit?: number;
+	offset?: number;
+	orderBy?: OrderBy<T> | OrderBy<T>[];
+	data?: Partial<T>;
+	upsertOnlyUpdateData?: Partial<T>;
+};
+
+/**
+ * @typeParam T - The type of the object to query. This is generally not needed to be specified, but can be useful if you're calling this yourself instead of through a {@link Model}.
+ * ```ts
+ * {
+ * 	// Any of these are valid
+ * 	orderBy: 'id',
+ * 	orderBy: ['id', 'name'],
+ * 	orderBy: { column: 'id', descending: true, nullLast: true },
+ * 	orderBy: [{ column: 'id', descending: true, nullLast: true }, { column: 'name', descending: false, nullLast: false }],
+ * }
+ * ```
+ */
+export type OrderBy<T extends object> =
+	| keyof T
+	| { column: keyof T; descending: boolean; nullLast?: boolean };
+
+/**
+ * @param type - The type of query to generate, see {@link QueryType}
+ * @param tableName - The table to query
+ * @param options - The options for the query, see {@link GenerateQueryOptions}
+ * @typeParam T - The type of the object to query. This is generally not needed to be specified, but can be useful if you're calling this yourself instead of through a {@link Model}.
+ * @returns The query and bindings to be executed
+ */
+export function GenerateQuery<T extends object>(
+	type: QueryType,
+	tableName: string,
+	options: GenerateQueryOptions<T> = {},
+	primaryKey = "id"
+): { bindings: unknown[]; query: string } {
+	if (typeof tableName !== "string" || !tableName.length) {
+		throw new Error("Invalid table name");
+	}
+	let query = "";
+	const bindings: unknown[] = [];
+	switch (type) {
+		case QueryType.SELECT: {
+			query = `SELECT * FROM ${tableName}`;
+			if (options.where) {
+				const whereStmt = [];
+				for (const [key, value] of Object.entries(options.where)) {
+					whereStmt.push(`${key} = ?`);
+					bindings.push(value);
+				}
+				query += ` WHERE ${whereStmt.join(" AND ")}`;
+			}
+			if (options.orderBy) {
+				query += " ORDER BY " + transformOrderBy(options.orderBy);
+			}
+			if (options.limit) {
+				query += ` LIMIT ${options.limit}`;
+				if (options.offset) {
+					query += ` OFFSET ${options.offset}`;
+				}
+			}
+			break;
+		}
+		case QueryType.DELETE: {
+			query = `DELETE FROM ${tableName}`;
+			if (options.where) {
+				const whereStmt = [];
+				for (const [key, value] of Object.entries(options.where)) {
+					whereStmt.push(`${key} = ?`);
+					bindings.push(value);
+				}
+				query += ` WHERE ${whereStmt.join(" AND ")}`;
+			}
+			break;
+		}
+		case QueryType.INSERT: {
+			query = `INSERT INTO ${tableName}`;
+			if (typeof options.data !== "object") {
```

Comment:
I'd also check here if the object has any keys, since doing INSERT with `{}` will fail.